index.js

```diff
@@ -185,4 +185,56 @@ document.getElementById('btn4').addEventListener('click', openDetails4);
 document.querySelector('.popup-cancel-icon1').addEventListener('click', closeDetail1);
 document.querySelector('.popup-cancel-icon2').addEventListener('click', closeDetail2);
 document.querySelector('.popup-cancel-icon3').addEventListener('click', closeDetail3);
-document.querySelector('.popup-cancel-icon4').addEventListener('click', closeDetail4);
\ No newline at end of file
+document.querySelector('.popup-cancel-icon4').addEventListener('click', closeDetail4);
+
+// form
+
+const formControl = document.getElementById('form');
+const userName = document.getElementById('user_name');
+const userEmail = document.getElementById('user_email');
+const userMessage = document.getElementById('message');
+const errorElement = document.getElementById('error');
+
+const validateForm = () => {
+  const emailVal = userEmail.value.trim();
+  if (/^[a-z0-9]+@[a-z-0-9]+\.[a-z0-9-.]+$/.test(emailVal)) {
+    errorElement.classList.remove('error');
+    errorElement.classList.add('success');
+    errorElement.innerHTML = 'Success';
+    formControl.submit();
+  } else {
+    errorElement.classList.remove('success');
+    errorElement.classList.add('error');
+    errorElement.innerHTML = 'email should be in lowercase';
+  }
+};
+
+formControl.addEventListener('submit', (event) => {
+  event.preventDefault();
+  validateForm();
+});
+
+// Local Storage (cache)
+
+function setFormData() {
+  userName.value = localStorage.getItem('userName');
+  userEmail.value = localStorage.getItem('userEmail');
+  userMessage.value = localStorage.getItem('userMessage');
+}
+
+function saveFormData() {
+  localStorage.setItem('userName', userName.value);
+  localStorage.setItem('userEmail', userEmail.value);
+  localStorage.setItem('userMessage', userMessage.value);
+  setFormData();
+}
```

Comment:
- [ ] Great job 👍 saving data to and retrieving data from `localStorage` ☺️  however it would be better 💯  if you save data into one single object like e.g.

```sh

const formData = {
    name: nameInput.value,
    email: emailInput.value,
    body: bodyInput.value,
};

```
and store it into the `localStorage`, instead of storing multiple objects for each field

As it is the [requirement](https://github.com/microverseinc/curriculum-html-css/blob/main/portfolio/10_localstorage.md#project-requirements) of the project:

> You must create a single JavaScript object with all the data from the entire form and save it in local storage. Do not create one local storage entry per input field.